Prevent the re-requesting of txns we already have

ptschip opened the pull request:
We were indicating that we had the txn when it was received however
if INV's were arriving after then we would end up re-requesting them
becuase the item had already been cleaned up from the requestmanager
queue. To fix this we just need to indicate that we received the txn
when it was either submitted to the mempool or the orphanpool. This
works because the request manager checks whether we have a txn in
either of those pools before requesting.

This is the same approach we have for blocks, where we don't indicate
we have the block until it is stored fully on disk and the blockindex
updated.

ptschip committed:
Prevent the re-requesting of txns

We were indicating that we had the txn when it was received however
if INV's were arriving after then we would end up re-requesting them
becuase the item had already been cleaned up from the requestmanager
queue. To fix this we just need to indicate that we received the txn
when it was either submitted to the mempool or the orphanpool. This
works because the request manager checks whether we have a txn in
either of those pools before requesting.

This is the same approach we have for blocks, where we don't indicate
we have the block until it is stored fully on disk and the blockindex
updated.

ptschip committed:
Update the transaction response time

gandrewstone commented:
I like this approach but am worried because the txCommitQ is flushed every 2 seconds by default, but we might even adjust that to 5 or above for greater parallelism.  So we may rerequest an INV even though we already have it during processing. 

Let me propose we add a flag to CUnknownObj meaning we've received data but are processing: "bool pending".  Then set pending to true in your UpdateTxnResponseTime(), and in SendRequests skip the rerequest if "pending" is true then this would solve the problem.  If during processing we find that the data is bogus -- that is, the received TX bytes' hash does not match the INV hash then we'd clear the pending flag, remove that node as a source, and give that node a "misbehaving".  This way we can recover from bad data.  If the data is not bad, the CUnknownObj gets removed during tx commitment as you have done and so we leave the "pending" state by deleting the object from the requestmanager.

ptschip committed:
Change where we mark the txn as received.

We have to mark it as received after processing even if it was
and invalid txn. This way the item will get cleaned up properly
in all instances.

ptschip commented:
@gandrewstone I like your idea. I think it's the way to go long term, but given the short time we have to get this release out I think it would be good to if we could merge this PR and fix the bug.  Then I can take a little more time and work your design in.  I like the idea of yours as to this tracking  state ...we might also want have some sort of processing timeout which could be much longer so that some txn didn't get stuck in a queue somewhere...

gandrewstone commented on src/requestManager.cpp:
```diff
@@ -257,36 +257,47 @@ bool CRequestManager::AlreadyAskedForBlock(const uint256 &hash)
+        cleanup(item);
     }
-    else if ((obj.type == MSG_BLOCK) || (obj.type == MSG_THINBLOCK) || (obj.type == MSG_XTHINBLOCK))
+    else if (pfrom && (obj.type == MSG_BLOCK || obj.type == MSG_THINBLOCK || obj.type == MSG_XTHINBLOCK))
```
If the block is received we should not skip this logic because we don't know what node it was received from, especially since the only use of pfrom is in a log.  Let's put an inline if in the log instead: pfrom ? pfrom->GetLogName() : "unknown"

gandrewstone commented:
I'm fine with deferring my idea.  I just have one more comment to be addressed, then will merge.

ptschip committed:
Return with "unknown" peer if no peer supplied

In requester.Received() if we supply a null for a peer when we
handle a block that was received then print out "unknown" for
the peers name.

ptschip commented:
done

On 27/09/2018 5:33 AM, Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/requestManager.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1357#discussion_r220904367>:
>
> >      }
> -    else if ((obj.type == MSG_BLOCK) || (obj.type == MSG_THINBLOCK) || (obj.type == MSG_XTHINBLOCK))
> +    else if (pfrom && (obj.type == MSG_BLOCK || obj.type == MSG_THINBLOCK || obj.type == MSG_XTHINBLOCK))
>
> If the block is received we should not skip this logic because we
> don't know what node it was received from, especially since the only
> use of pfrom is in a log. Let's put an inline if in the log instead:
> pfrom ? pfrom->GetLogName() : "unknown"
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1357#pullrequestreview-159421241>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0ITZMd2tiOc7sWHmjQEbVFfKdTwJks5ufMWsgaJpZM4W7Egp>.
>
